Consistently/correctly handle numpy integral types

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30486 Correctly handle scalar types, fix parse of numpy ints**

Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some inconsistent/incorrect code in the handling of parsing python/numpy scalars that led to issue #29252:

Where we allow interpretation of a zero-dim numpy integer value as a scalar in pytorch, we incorrectly parsed the int as a float.

Also fixes the issue described here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

And also allows numpy integral types to be used in cases where we allow python ints (treated as long internally).

Test Plan:

Added a unit test based on the example given in the issue, a test for numpy ints being used as ints in operator arguments, and a couple of miscelaneous test cases.

Differential Revision: [D18932520](https://our.internmc.facebook.com/intern/diff/D18932520)

nairbv committed:
Correctly handle scalar types, fix parse of numpy ints

Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some inconsistency in the code for parsing python numbers, e.g.
a function named THPUtils_checkDouble that returns true for integral
values as well as floating point values, and incorrect code in
python_arg_parser.

This led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some inconsistency in the code for parsing python numbers, e.g.
a function named THPUtils_checkDouble that returns true for integral
values as well as floating point values, and incorrect code in
python_arg_parser.

This led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some inconsistency in the code for parsing python numbers, e.g.
a function named THPUtils_checkDouble that returns true for integral
values as well as floating point values, and incorrect code in
python_arg_parser.

This led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some inconsistency in the code for parsing python numbers, e.g.
a function named THPUtils_checkDouble that returns true for integral
values as well as floating point values, and incorrect code in
python_arg_parser.

This led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

mruberry commented:
Great to see that test fixed, but @gchanan is (or knows) a better reviewer.

zou3519 commented on test/test_torch.py:
```diff
@@ -5221,6 +5221,15 @@ def test_multiplication_numpy_scalar(self):
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_parse_numpy_int(self):
+        # https://github.com/pytorch/pytorch/issues/29252
+        np_val = np.ones(1, dtype=np.int64)[0]
```
nit: Might be clearer as ` np.dtype('int32').type(1)`

nairbv replied:
I find that less clear, partly because it isn't directly translatable to pytorch

zou3519 commented on test/test_torch.py:
```diff
@@ -5221,6 +5221,15 @@ def test_multiplication_numpy_scalar(self):
+        # https://github.com/pytorch/pytorch/issues/29252
+        np_val = np.ones(1, dtype=np.int64)[0]
+        self.assertEqual(type(np_val), np.int64)
+        th_val = torch.ones(size=(1,), dtype=torch.long)
```
nit: If you just wanted a scalar you can do `torch.ones([], dtype=torch.long)`

nairbv replied:
I do want 1 dim not zero-dim though here

zou3519 replied:
Why?

nairbv replied:
You said "If you just wanted a scalar," but this isn't a scalar. I'm not exactly sure what you mean.

The example came from the specific reported bug.  I wrote the test first based on the given example, verified that the test failed, then made the changes to make the test pass.

zou3519 replied:
> The example came from the specific reported bug. I wrote the test first based on the given example, verified that the test failed, then made the changes to make the test pass.

Thanks, that's the explanation I was looking for

zou3519 commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -138,13 +138,21 @@ bool FunctionParameter::check(PyObject* obj) {
       // fallthrough
     case ParameterType::DOUBLE: {
-      if (THPUtils_checkDouble(obj)) {
+      if (THPUtils_checkDouble(obj) || THPUtils_checkLong(obj)) {
```
I think the problem was that `THPUtils_checkDouble` doesn't mean "check that the input is a double type", but rather, "check that the input is cast-able to double type".

nairbv replied:
yes that ambiguity is part of the issue. I might be able to revert the meaning to the original 'castable' one after having fixed issues elsewhere though. I'll check.

zou3519 commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -654,9 +662,9 @@ at::Tensor PythonArgs::tensor_slow(int i) {
+  } else if (PyComplex_Check(obj)) {
     scalar = at::Scalar(THPUtils_unpackComplexDouble(obj));
-  }else if (THPUtils_checkDouble(obj)) {
+  } else if (THPUtils_checkDouble(obj)) {
```
There's a callsite in tuple_parser.cpp that we might have to change. I'm not sure what it does though but it would be good to figure out what it is doing and how this PR affects it:
https://github.com/pytorch/pytorch/blob/e6000a7c045cbece5fbfd7d933c39e40b1625037/torch/csrc/utils/tuple_parser.cpp#L42

nairbv replied:
hmmm... I think you're probably right but I'm not seeing any calls to this function that I could use to validate expected behavior. Do you think it would be alright to just remove tuple_parser?

I see one use for string parameters in torch/csrc/autograd/functions/init.cpp, but I think a small modification there would allow me to eliminate the rest of this tuple_parser code.

nairbv replied:
for now I'm reverting the change to that function, which (assuming the tests pass this way) bypasses the issue. I'm still tempted to remove tuple_parser in a subsequent PR though if that makes sense.

zou3519 commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -41,10 +41,15 @@ inline PyObject* THPUtils_packDoubleAsInt(double value) {
 inline bool THPUtils_checkLong(PyObject* obj) {
+  bool is_numpy_int = false;
+#ifdef USE_NUMPY
+  is_numpy_int = torch::utils::is_numpy_int(obj);
```
could we just do
```
if(torch::utils::is_numpy_int(obj)) [
    return true
}
```
instead of adding the "|| is_numpy_int" condition below?

nairbv replied:
yes that'll probably be clearer

zou3519 commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -227,6 +230,10 @@ ScalarType numpy_dtype_to_aten(int dtype) {
       ((PyTypeObject*)pytype.get())->tp_name);
 }
 
+bool is_numpy_int(PyObject* obj) {
```
For np.bool:
- `is_numpy_int` returns false
- `is_numpy_scalar` returns true

Is that correct?

nairbv replied:
yes

zou3519 replied:
I stepped through this in gdb. `PyArray_IsIntegerScalar` returns False for np.bool, so we don't need the extra check for bool. The relevant source code is here: https://github.com/numpy/numpy/blob/4bc7eda1f0ff87a9e6eaace60ef66b6c1b9658a1/numpy/core/include/numpy/ndarrayobject.h#L61-L63 but it's a little hard to read.

In addition, we could clean up `is_numpy_scalar` by using `is_numpy_int`.

nairbv replied:
If I'm understanding correctly, that seems to imply that all of the existing code like `return (PyLong_Check(obj) || PyInt_Check(obj)) && !PyBool_Check(obj);` is also incorrect? 

PyArray_IsIntegerScalar etc appears to rely on the same internal interpretation of looking for Py_TPFLAGS_LONG_SUBCLASS on the type object's tp_flags, which seems isn't set for bool types.

zou3519 replied:
> If I'm understanding correctly, that seems to imply that all of the existing code like `return (PyLong_Check(obj) || PyInt_Check(obj)) && !PyBool_Check(obj);` is also incorrect?

No, the existing code is correct because PyLong_Check returns True for Python booleans.

`PyArray_IsIntegerScalar` returns False for np.bool, but True for PyBool, which is kinda fun. So given that, my question is, what should `is_numpy_int` actually check? What does it return for np.int64, PyLong, np.bool, PyBool, np.double, PyFloat?

nairbv replied:
hmm... It seems PyLong_Check effectively just checks `((t)->tp_flags & Py_TPFLAGS_LONG_SUBCLASS) != 0`, where Py_TPFLAGS_LONG_SUBCLASS == 1 << 24. Is that not right?

It appears that np.bool and bool should behave the same?

```
In [14]: bool.__flags__ & (1 << 24)
Out[14]: 16777216
In [15]: np.bool.__flags__ & (1 << 24)
Out[15]: 16777216
```
What am I misunderstanding about PyLong_Check?

zou3519 replied:
I'm inferring the behavior of PyLong_Check from the documentation. I haven't actually run any code or read the implementation of PyLong_Check.

Based on the documentation:
- The documentation says "Return true if its argument is a PyLongObject or a subtype of PyLongObject.": https://docs.python.org/3/c-api/long.html#c.PyLong_Check.
- Then, in the documentation for boolean, "Booleans in Python are implemented as a subclass of integers." https://docs.python.org/2/c-api/bool.html

What happens empirically if you write a program that checks the value of `PyLong_check(Py_True)`?

zou3519 replied:
It looks like there's a difference between np.bool and np.bool_: https://joergdietrich.github.io/python-numpy-bool-types.html#Numpy-bool. `np.bool` is the same thing as Python bool, while `np.bool_` is what you get if you create a bool array in numpy

nairbv replied:
interesting... 
```
In [10]: np.array([True]).dtype.type
Out[10]: numpy.bool_
In [11]: np.array([True]).dtype.type.__flags__ & (1 << 24 )
Out[11]: 0
```

zou3519 commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -100,25 +105,17 @@ inline bool THPUtils_unpackBool(PyObject* obj) {
-  return PyFloat_Check(obj) || PyLong_Check(obj) || PyInt_Check(obj) || is_numpy_scalar;
-#else
-  return PyFloat_Check(obj) || PyLong_Check(obj) || is_numpy_scalar;
+  is_numpy_double = torch::utils::is_numpy_scalar(obj) && ! torch::utils::is_numpy_int(obj);
```
What does this do for `np.bool`? Is this true?

nairbv replied:
np.bool returns false, consistent with the original implementation of checkLong above which for python objects is `PyLong_Check(obj) && !PyBool_Check(obj)`

zou3519 replied:
So np.int64 is not a numpy double, but np.bool is a numpy double for the purposes of this function? The naming seems off and I wonder what consequences this has

nairbv replied:
reverted the behavioral changes back to the meaning of "castable," so now it's `scalar && ! bool`.

zou3519 reviewed (commented):
Some initial questions

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some inconsistency in the code for parsing python numbers, e.g.
a function named THPUtils_checkDouble that returns true for integral
values as well as floating point values, and incorrect code in
python_arg_parser.

This led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some inconsistency in the code for parsing python numbers, e.g.
a function named THPUtils_checkDouble that returns true for integral
values as well as floating point values, and incorrect code in
python_arg_parser.

This led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -100,30 +106,28 @@ inline bool THPUtils_unpackBool(PyObject* obj) {
-  is_numpy_scalar = torch::utils::is_numpy_scalar(obj);
-#else
-  is_numpy_scalar = false;
+  if (torch::utils::is_numpy_scalar(obj)) {
```
I didn't change the behavior for checkScalar (since it wasn't _directly_ related to the reported bug), but it looks like the existing code would return true for a numpy boolean where it probably shouldn't (the below doesn't seem to include boolean).

I'm not very familiar with how we expect complex types to work in pytorch.

zou3519 replied:
I checked that `is_numpy_scalar` returns false for np.bool

zou3519 commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -100,30 +106,28 @@ inline bool THPUtils_unpackBool(PyObject* obj) {
-  is_numpy_scalar = torch::utils::is_numpy_scalar(obj);
-#else
-  is_numpy_scalar = false;
+  if (torch::utils::is_numpy_scalar(obj) && ! PyBool_Check(obj)) {
```
Why did this change?

zou3519 replied:
On master,
```
tensor = torch.zeros(3)
tensor.bernoulli_(p=True)
```
works. `p` is a `float` argument. I'm not sure if this is intentional or not, but the check against PyBool here would make it so that `tensor.bernoulli_(p=True)` errors out.

nairbv replied:
I think you're right it's probably not needed and so I'll remove it, but I think it would only potentially affect something like `tensor.bernoulli_(p=np.array([True])[0])` which doesn't work on master either.

zou3519 replied:
It doesn't affect `tensor.bernoulli_(p=np.array([True])[0])` -- `PyBool_Check` returns False for np.bool_ so having it there is redundant.

nairbv replied:
right (and removed)

zou3519 commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -148,6 +148,15 @@ std::vector<int64_t> compute_sizes(PyObject* seq) {
 }
 
 ScalarType infer_scalar_type(PyObject *obj) {
+#ifdef USE_NUMPY
```
Why did we move this code block?

nairbv replied:
Functions below like THPUtils_checkLong() are checking if it's a numpy long or python long. If we run the numpy checks last, we'll never reach the numpy checks so won't check the numpy dtype.

zou3519 replied:
If `THPUtils_checkLong(obj)` decides that obj is a numpy long, then `infer_scalar_type` returns `ScalarType::Long`. This seems correct to me -- what is wrong there?

nairbv replied:
PyArray_CheckScalar will return true for a np.int32 scalar, and get the correct torch.int32 result type. If we rely on THPUtils_checkLong we always return Long.

zou3519 replied:
I'm confused. The test https://github.com/pytorch/pytorch/pull/30486/files/12d6de600476b9b1d9bcae6351680b37995e6d9c#diff-f539198dd366265fb8dc2d661bc5d5bcL1507 seems to test that a np.int32 gets converted to a torch.IntTensor, which worked before this PR. Is the test wrong?

nairbv replied:
Yes the test was semi-broken on master (and fixed by this PR).

On master, THPUtils_checkDouble checks for numpy scalars, but THPUtils_checkLong doesn't. This PR changes checkLong  to also check for numpy ints, consistent with checkDouble. On master the test was broken for float/double depending on the result of get_default_scalar_type() below. (the actual issue on master was a short-circuiting for np.float/np.double not for int32, though the same issue gets exposed for int32 as well once changing other things in this PR).

See here for a related change:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

zou3519 replied:
Pitch: Instead of moving the codeblock, we should change the call to `THPUtils_checkLong` below to a call to a new function `THPUtils_checkPythonLong` (or introduce a `THPUtils_checkLong(PyObject*, bool include_numpy_long=True)`)

Before this pull request, all of the cases are exclusive; none of them overlap. After this pull request, some of the checks overlap. This is bad for code readability: "Does THPUtils_checkLong mean check that the value is convertible to a long, or is it checking for a python long?"

Making the checks exclusive also means that we don't have to worry about the order in which they appear, like we are doing now. With this pull request, there is a non-trivial reason things in this function are ordered the way they are.

nairbv replied:
> Before this pull request, all of the cases are exclusive

I don't think that's true. Before this PR, a numpy float scalar was being caught by one these checks and short circuiting to return `get_default_dtype` without reaching the numpy handling block (which it also would have satisfied). To make this mutually exclusive, we'd also need something like maybe a wrapper function to check `PyFloat_Check and not numpy`. To me, that seems like overkill, at least for purposes of this bug-fix.

We can keep iterating after this lands too if we can come up with a better code structure.

zou3519 replied:
Which case does a `numpy.float64` scalar go into?

zou3519 replied:
Oh I see, PyFloat_check returns True for numpy.float64... On the other hand, PyLong_check returns False for numpy.int64 :/

We could use PyFloat_CheckExact instead of PyFloat_Check. But if you think this is overkill, we should at least include a comment in the code explaining that the numpy checks must go first otherwise they'll be considered to act like a python float or long.

zou3519 reviewed (changes requested):
Some more questions

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

zou3519 commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -41,6 +41,12 @@ inline PyObject* THPUtils_packDoubleAsInt(double value) {
 
 inline bool THPUtils_checkLong(PyObject* obj) {
+#ifdef USE_NUMPY
+  if (torch::utils::is_numpy_int(obj)) {
```
I'm a little concerned about changing the behavior of THPUtils_checkLong because it has a lot of call sites. I'm worried that it enables new behavior in a couple of places and we should figure out if we want that new behavior and add tests for them if so.

nairbv replied:
That's a good point. For a couple of quick examples, I can see that this branch introduces the ability to: `torch.Storage(np.array([1])[0])`, where on master I get a TypeError expecting an int.  I also found I'm able to `torch.ones(5)[0]=np.array([1])[0]` on this branch but not on master.

The usage seems valid, and doesn't break any existing behavior. It allows treating `np.array([an_int])[0]` the same as `an_int` in many cases where `an_int` was considered valid int input.

I can add these couple of examples. There could be a lot of individual cases though, and I'm not sure it's worth the investment in auditing and testing every possible case.

zou3519 replied:
> I can add these couple of examples. There could be a lot of individual cases though, and I'm not sure it's worth the investment in auditing and testing every possible case.

How many callsites are there? If any of those cases crashes then we have ourselves multiple high-pri bugs in the future.

zou3519 replied:
We can also now accept np.int32 as a Long type, which is probably fine

nairbv replied:
```
~/repos/pytorch]$ grep -Irn THPUtils_checkLong . | wc -l
63
```
Many of these are in C++ functions that back python functions which start with an `_`, e.g., THPModule_setQEngine / `_set_qengine`. We'd have to fix it if someone passed a numpy scalar and it somehow crashed, but that seems unlikely, and also not like a particularly egregious/risky potential  bug.

zou3519 replied:
We don't have to *test* every possible case, but we should at least understand the consequences of our change.

nairbv replied:
I don't think anyone would be that surprised if stuff like this crashed for a numpy int, already crashes for `1`:

```
In [2]: torch._C._set_qengine(1)
libc++abi.dylib: terminating with uncaught exception of type c10::Error: quantized engine FBGEMM is not supported (setQEngine at ../aten/src/ATen/Context.cpp:113)
```

nairbv replied:
It seems like anywhere int works and uses this function, numpy int should also now work. testing the first couple of cases I could find this seems to be the case (and added unit tests for a couple of the more likely-to-be-used cases).

zou3519 replied:
I looked into this a little more. Almost every use case of THPUtils_checkLong is followed by a `THPUtils_unpackLong`. As long as `THPUtils_unpackLong` supports all of the new types we are allowing in (np.(u)int64, np.(u)int32, np.(u)int16, np.(u)int8) then the callsites should be fine, but I would like to see testing for these types.

The other question then is the UI question. Should we be accepting all of these types as "Long"?

nairbv replied:
Yes that was what I saw as well. I don't see any reason to reject a smaller or unsigned scalar when we're expecting a python int. I modified the test to check for a few np dtypes (though I didn't include uint32/64 since we don't have an unsigned 32/64 bit integral type in torch).

zou3519 replied:
So what actually happens if we try to use an unsigned numpy type with pytorch?

nairbv replied:
it's treated as a long just like everything else

zou3519 replied:
Does np.uint64 overflow if we pass it in for a parameter like `torch.sum(x, dim=???)`? Does it error out gracefully?

nairbv replied:
do you know of an example I could try? I don't think we can create a tensor with uint64.max > dims > int64.max

It should behave just like any int since `THPUtils_unpackLong` will do the same overflow checking it does anyway (python ints can be larger than a long)

zou3519 replied:
You don't need to create a tensor with that many dimensions to test if `torch.mean(torch.randn(1, 1), np.uint64(-1))` errors out gracefully or not. 

I don't know what PyLong_AsLongLongAndOverflow does and/or if it handles np.uint64. PyLong should only work on PyLong and subclasses of it (the np integer types aren't actually subclasses of PyLong...) so we should test the behavior.

nairbv replied:
I think the np integer types are subclasses of PyLong (by setting the Py_TPFLAGS_LONG_SUBCLASS type flag in their custom type).

Checking, `torch.mean(torch.randn(1, 1), np.uint64(-1))` gives `RuntimeError: Overflow when unpacking long`
on this branch as I'd expect. This does differ from master which gives `TypeError: mean() received an invalid combination of arguments - got (Tensor, numpy.uint64), but expected one of:...`

zou3519 replied:
Sounds good, can you add that as a test case?

nairbv replied:
added

zou3519 reviewed (commented):
We should probably fix this comment: https://github.com/pytorch/pytorch/blob/f2a2fec47cd7f3b715189338e45c6aebdf5f65f2/torch/csrc/utils/python_arg_parser.cpp#L766-L767 which suggests that all numpy types other than np.float64 aren't handled correctly

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

zou3519 commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -214,6 +214,14 @@ auto FunctionParameter::check(PyObject* obj, std::vector<py::handle> &overloaded
       return allow_numbers_as_tensors && THPUtils_checkScalar(obj);
     }
     case ParameterType::SCALAR:
+      if (THPUtils_checkScalar(obj)) {
```
Why do we need the extra code here? Is it for readability? The original code does a fallthrough to ParameterType::COMPLEX and then ParameterType::DOUBLE; all of this code seems like it is covered in those two fallbacks

nairbv replied:
It looks like the original idea was for it to fall through, but no that doesn't work. If we reach `ParameterType::Scalar` but the value is an int, we'll fall through to `case ParameterType::Double`, fail the THPUtils_checkDouble, and return false without reaching the `case ParameterType::INT64` case.

zou3519 replied:
I don't think that's correct -- `THPUtils_checkDouble` accepts ints: https://github.com/pytorch/pytorch/blob/bf1b4b6feff7be16df2473a54931569465ce0ffc/torch/csrc/utils/python_numbers.h#L102-L114

nairbv replied:
I can try again but I seem to remember this not having worked correctly without the change to prevent falling through to a `return false`. Maybe it was bool or maybe it's not still the case after other changes.

nairbv replied:
I'm not loving this, but it does seem to work.

zou3519 commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -227,9 +230,13 @@ ScalarType numpy_dtype_to_aten(int dtype) {
 bool is_numpy_scalar(PyObject* obj) {
-  return (PyArray_IsIntegerScalar(obj) ||
-          PyArray_IsScalar(obj, Floating));
+  return is_numpy_int(obj) ||
```
nit: This all fits onto one line so we don't need to split it

zou3519 commented on test/test_torch.py:
```diff
@@ -5260,6 +5260,21 @@ def test_multiplication_numpy_scalar(self):
+        self.assertEqual((np_val + th_val).dtype, torch.long)
+        self.assertEqual((th_val + np_val).dtype, torch.long)
+
+        # validate use of np.int in some other cases that accept python int
```
np.int is an alias for the python built-in int type. We should call this `np.int64`

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -5260,6 +5260,22 @@ def test_multiplication_numpy_scalar(self):
+        # https://github.com/pytorch/pytorch/issues/29252
+        for nptype in [np.int16, np.int8, np.uint8, np.int32, np.int64]:
+            np_arr = np.array([5], dtype=nptype)
+            np_val = np_arr[0]
```
nit: all of these can be written as `np_val = np.int64(5)`

nairbv replied:
I want the np_arr below to call torch.from_numpy

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -107,7 +107,7 @@ inline bool THPUtils_unpackBool(PyObject* obj) {
 inline bool THPUtils_checkDouble(PyObject* obj) {
 #ifdef USE_NUMPY
-  if (torch::utils::is_numpy_scalar(obj) && ! PyBool_Check(obj)) {
+  if (torch::utils::is_numpy_scalar(obj)) {
```
removing the bool check here caused test failures:

```
FAIL: test_default_collate_dtype (__main__.TestDataLoader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bvaughan/repos/pytorch/test/test_dataloader.py", line 1531, in test_default_collate_dtype
    self.assertEqual(collated.dtype, torch.bool)
  File "/home/bvaughan/repos/pytorch/test/common_utils.py", line 841, in assertEqual
    super(TestCase, self).assertEqual(x, y, message)
AssertionError: torch.int64 != torch.bool :
```

zou3519 replied:
On master we don't have the bool check: https://github.com/pytorch/pytorch/blob/244b0bd1a5a64da370f44fca7ba4a73e296dd4b0/torch/csrc/utils/python_numbers.h#L102-L114 , so why is the bool check necessary now?

nairbv replied:
sorry I think I commented on the wrong line. it looks like the tests were broken where we removed the PyBool_Check in is_numpy_int, in this same commit, (tensor_numpy.cpp).

Master is wrong though, so we can't rely on what master did to tell us what's correct.

zou3519 replied:
> Master is wrong though, so we can’t rely on what master did to tell us what’s correct 

Master isn’t completely wrong because the code works for applications other than the use case this PR is trying to fix. I’m not suggesting that master is correct on this; I’m probing about why we are doing things to better understand what is going on

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -231,12 +231,12 @@ ScalarType numpy_dtype_to_aten(int dtype) {
 
 bool is_numpy_int(PyObject* obj) {
-  return PyArray_IsIntegerScalar(obj) && ! PyBool_Check(obj);
+  return PyArray_IsIntegerScalar(obj);
```
this one

zou3519 commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -227,9 +230,12 @@ ScalarType numpy_dtype_to_aten(int dtype) {
 }
 
+bool is_numpy_int(PyObject* obj) {
+  return PyArray_IsIntegerScalar(obj) && ! PyBool_Check(obj);
```
So the question I have right now is: what is the root reason we have to guard against bool here? I understand that the tests pass with it and they don't without it, but that doesn't say why.

nairbv replied:
This `is_numpy_int` function was added to add numpy support for THPUtils_checkLong, and that's really the only reason it's here. The implementation for checkLong on master was `return PyLong_Check(obj) && !PyBool_Check(obj);`. 

If we allow checkLong to treat booleans as long, other code breaks. It's hard to track down everything it would break, but presumably it's because of one of these if/else statements checking types of parameters.

zou3519 replied:
Does PyArray_IsIntegerScalar return True for Py_Bool?

zou3519 replied:
It looks like the answer to that is Yes, because it calls PyLong_Check, which returns True for bool. 

if you’re checking that something is an np int then the best api to use looks like
 https://github.com/numpy/numpy/blob/1783e6fc437f34af1d652cbae36a68abcbf41fbc/numpy/core/include/numpy/ndarrayobject.h#L53 , not PyArray_IsIntegerScalar.

nairbv replied:
> Does PyArray_IsIntegerScalar return True for Py_Bool?

Yes because `bool.__flags__ & (1 << 24) != 0` as we looked at before.

FYI, on master, related to the np.bool vs np.bool_ that you pointed out, I'm not trying to change this particular behavior in this PR:

```
>>> torch.pow(torch.ones(4), True)
tensor([1., 1., 1., 1.])
>>> torch.pow(torch.ones(4), np.array([True])[0])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: pow() received an invalid combination of arguments - got (Tensor, numpy.bool_), but expected one of:
```

zou3519 replied:
> FYI, on master, related to the np.bool vs np.bool_ that you pointed out, I'm not trying to change this particular behavior in this PR:

I agree that we're not trying to change this behavior. In the np.bool vs np.bool_ thread we had going, I asked what the behavior of this function was on np.bool_.

I think that instead of writing `return PyArray_IsIntegerScalar(obj) && ! PyBool_Check(obj);` we should write return `PyArray_IsScalar((obj), Integer))`, like in here: https://github.com/numpy/numpy/blob/1783e6fc437f34af1d652cbae36a68abcbf41fbc/numpy/core/include/numpy/ndarrayobject.h#L52-L53

Otherwise, the semantics of `is_numpy_int` become too confusing: `is_numpy_int(PyBool)` is False right now, but `is_numpy_int(PyLong)` is True. I think what we really want is:
- is_numpy_int(Python number type) -> False
- is_numpy_int(numpy integer type) -> True

nairbv replied:
ah, yes that makes sense, I'll try that.

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -5260,6 +5260,22 @@ def test_multiplication_numpy_scalar(self):
+            self.assertEqual(tensor[0], np_val)
+
+            t = torch.from_numpy(np_arr)
+            self.assertEqual((t + np_val).dtype, (t + t[0].item()).dtype)
```
Isn't `(t + t[0].item()).dtype` just `t.dtype` ? 

What exactly is this testing?

nairbv replied:
ah, yeah this isn't necessary and I'll fix it. t[0].item() will always be long or double, so it is "special" in a way, but within a type category it won't ever actually change the type of the result since it's being added to a tensor with dims.

nairbv replied:
fixed

zou3519 commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -148,6 +148,15 @@ std::vector<int64_t> compute_sizes(PyObject* seq) {
+    THPObjectPtr arr(PyArray_FromScalar(obj, nullptr));
+    return numpy_dtype_to_aten(PyArray_TYPE((PyArrayObject*) arr.get()));
+  }
+#endif
```
BTW, while you're modifying this file, can you make `infer_scalar_type` static? There is another function called `infer_scalar_type` in the codebase and I think that they're distinct.

zou3519 replied:
NVM, I see that it is in an anonymous namespace, so no need to make it static.

nairbv replied:
updated

nairbv replied:
oops

zou3519 reviewed (commented):
Implementation looks great. I'm a bit confused about the testing

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

zou3519 commented:
I think we should update the title and/or body to say something along the lines of "we are now accepting np integral types for any Long arguments".

Along the lines of that, here are the last things I'm looking at:
- I'm still wondering if it is OK for us to accept things like np.int8 as Long. To make a decision I'm going to look at if numpy operators (np.mean) can accept np.int8 (my guess is yes, from our discussions, but I'd like to confirm it).
- I'm trying to think about how much testing is necessary for this change. Enabling np integral types in THPUtils_CheckLong means that we can accept np integral types in all functions that accept Long. However, it also does something(?) in the case of functions that take Scalar; I am not sure what exactly.

zou3519 commented on test/test_torch.py:
```diff
@@ -5260,6 +5260,24 @@ def test_multiplication_numpy_scalar(self):
+        for nptype in [np.int16, np.int8, np.uint8, np.int32, np.int64]:
+            np_arr = np.array([5], dtype=nptype)
+            np_val = np_arr[0]
+            self.assertEqual(torch.Storage(np_val).size(), 5)
```
Regarding testing the new behavior:

I think we want to test the following things:
1) A numpy int parses correctly as a Long (int64) argument
2) A numpy int parses correctly for a Scalar argument (that it gets parsed as int64)
3) Some misc coverage.

What is currently in this test:
a) construct torch.Storage(np_int)
b) Assign `tensor[0]` a np_int
c) Add a `tensor` and an `np_int`. 

Case 2 is covered by (c). Constructing the storage and assigning a value falls into case (3) (because those aren't native functions and they have custom logic in their python bindings). We should have something for case 1 (something like torch.mean, that takes a long argument).

It would be good to add comments demonstrating what we're testing here. It's not clear to me as a reader how Pytorch parses `t + np_val` works, so some context that this is testing the np_int -> Scalar (Long) route might be nice, so something like:
```
// Test parsing numpy integer as pytorch Scalar gives a pytorch Long
self.assertEqual((t + np_val).dtype, t.dtype)
self.assertEqual((np_val + t).dtype, t.dtype)
```

Finally, I'm not sure if this is necessary, but maybe we could have some extra tests to check that the pytorch behavior is the same as the numpy behavior

zou3519 reviewed (changes requested):
I checked np's behavior and they do accept their integer types for things like `np.mean`, so I think it is fine that we do that as well in this PR.

The last two action items I see are:
1) I had some comments on improving the testing / making the logic of the testing clearer
2) We should change the pull request title, body, and commit message to reflect that we're now adding new behavior: that we are now parsing all numpy integer types as Longs

nairbv committed:
Update on "Correctly handle scalar types, fix parse of numpy ints"


Fixes: https://github.com/pytorch/pytorch/issues/29252

Summary:

There is some incorrect code in the handling of parsing python numbers that led to issue #29252:

When we allow interpretation of a zero-dim numpy integer value
as a scalar in pytorch, we incorrectly parse the int as a float.

This PR also fixes the issue described in the "FIXME" here:
https://github.com/pytorch/pytorch/pull/27628/files#diff-f539198dd366265fb8dc2d661bc5d5bcR1487

Test Plan:

Added a unit test based on the example given in the issue.

[ghstack-poisoned]

zou3519 reviewed (approved):
Looks good but please change the PR title, body, and commit message to indicate that we are now enabling passing in all numpy int types as a pytorch Long.
